client/extensions/woocommerce/state/ui/shipping/zones/methods/reducer.js

```diff
@@ -38,15 +43,100 @@ reducer[ WOOCOMMERCE_SHIPPING_ZONE_METHOD_ADD ] = ( state, action ) => {
 	const id = nextBucketIndex( state.creates );
 	let method = { id, methodType, enabled: true };
 	if ( builtInShippingMethods[ methodType ] ) {
-		method = { ...method, ...builtInShippingMethods[ methodType ]( undefined, action ) };
+		method = {
+			...method,
+			...builtInShippingMethods[ methodType ]( undefined, action ),
+			title: getMethodName( methodType ),
+		};
 	}
 	return { ...state,
 		creates: [ ...state.creates, method ],
+		currentlyEditingId: id,
+		currentlyEditingChanges: { isNew: true },
+	};
+};
+
+reducer[ WOOCOMMERCE_SHIPPING_ZONE_METHOD_OPEN ] = ( state, action ) => {
+	return { ...state,
+		currentlyEditingId: action.methodId,
+		currentlyEditingChanges: {},
+	};
+};
+
+reducer[ WOOCOMMERCE_SHIPPING_ZONE_METHOD_CANCEL ] = ( state ) => {
+	return { ...state,
+		currentlyEditingId: null,
+	};
+};
+
+reducer[ WOOCOMMERCE_SHIPPING_ZONE_METHOD_CLOSE ] = ( state ) => {
+	const { currentlyEditingId, currentlyEditingChanges } = state;
+
+	if ( null === currentlyEditingId ) {
+		return state;
+	}
+	if ( isEmpty( currentlyEditingChanges ) ) {
+		// Nothing to save, no need to go through the rest of the algorithm
+		return { ...state,
+			currentlyEditingId: null,
+		};
+	}
+
+	const { changedType } = currentlyEditingChanges;
```

Comment:
Same comment as above, I think this "flag" should be outside of `currentlyEditingChanges`.